### DISCUSSION TITLE
[2023-05-08] GUI Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6596] [Weekly Demo Day]
## Summary 

I implemented the [Dynamic app resampling and better performance measurements PR](https://github.com/enso-org/enso/pull/6595).

<br/>

## Proper measurement of GPU performance.

**WARNING! Currently (2023) this extension is only available in Chrome Desktop.**

A new way to measure GPU performance have been introduced. It bases on the [ExtDisjointTimerQueryWebgl2 WebGL extension](https://developer.mozilla.org/en-US/docs/Web/API/EXT_disjoint_timer_query), which is pretty tricky to use. First of all, it was banned from all browsers some time ago because of the ["GLitch" exploit] (https://www.vusec.net/projects/glitch). However, as Site Isolation shipped in Chrome on Desktop, the extension was re-enabled there because the only data that could be read was data that came from the same origin. To learn more, see: https://bugs.chromium.org/p/chromium/issues/detail?id=1230926. 

<br/>

## Automatic low/high resolution modes.
The application resolution mode is automatically switched between low and hight. In the low resolution mode, the [device pixel ratio](https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio) will be set to `1.0` even on high-dpi screens. This will result in faster rendering (every 4-th pixel) and a more blurry image of the GUI. I believe that the blurriness of the fonts can be decreased in this mode by tweaking shaders, and it can be performed in a separate PR. The switch happens after a few frames were rendered slow/fast enough. The performance of the rendering is measured using the new performance measurements described above. In case the app is run not in Chrome, the FPS will be used instead of the proper measurements, which might cause the app to switch to low-resolution mode even if the bottleneck is CPU, not GPU.

https://user-images.githubusercontent.com/1623053/236707705-2379ac44-bcf0-47e5-a5f9-210f9ee5f6b1.mov

<br/>

## Improvements to the performance monitor.
The performance monitor was extended with additional information:
- The "frame time" was renamed to "GPU + CPU time".
- The "CPU time" is a new measurements which reports the time spent on CPU-related operations, excluding GPU ones.
- The "GPU time" is a new measurement which reports the time spent on GPU-related operations, excluding CPU ones.
- Idle time is a new measurement which reports how much free time was left during rendering of the frame. Please note, that currently this measurement is broken and it can report negative values, as we need to enable high-resolution timers in Enso.

Please note, that the division between CPU and GPU time is not known immediately and we can get it with a few frames delay. That's why you can observe sometimes in the performance monitor a greyed-out value for a few frames, which turns into a real value after some time:

https://user-images.githubusercontent.com/1623053/236707998-f47b7b5a-a4b3-47fe-9095-98114d02b57d.mov

**Comment by MichaelMauderer** [comment]
*  Disregard visualisation for selection of nodes. #6487 

https://user-images.githubusercontent.com/1428930/235451552-d99cf4c7-4420-4412-b553-e9bf6615a8a8.mp4
*  Fix layer ordering between dropdown and breadcrumbs backgrounds. #6483 
*  Implement loading spinner for visualisations. #6512 

https://user-images.githubusercontent.com/1428930/236801655-67a0ffed-da5d-4e27-8797-cd8126cb86d9.mp4
*  Fix visualisation type selector artifacts rendered after node preview visualisation was closed. #6575

**Comment by Procrat** [comment]
- #6530

  Before:
  
  https://user-images.githubusercontent.com/12892578/228822063-d752f742-0784-4b03-b19f-b4842a553a4d.mov
  
  After:
  
  https://user-images.githubusercontent.com/607786/235908932-0b579d69-5fd8-438e-a82b-02678d8e6156.mp4
  
- #6535 in the hopes that it might resolve an intermittent Wasm compilation bug we've been experiencing (#6091) which was significantly slowing down development time because it required a clean build every time.
- (I also started working on other issues, but nothing completed yet.)

**Comment by vitvakatu** [comment]
The fix for [Last named argument gets doubled when the first argument is removed#6228](https://github.com/enso-org/enso/issues/6228) was merged. (it was presented on the last demo)

Also, a very annoying issue #6411 was fixed after a long investigation.

At the moment I'm working on fixing #6485

**Comment by mwu-tow** [comment]
* Fixing the warning about the version, caused by release: https://github.com/enso-org/enso/pull/6570
* Looking into various CI issues, notably:
  * Lint job problems in GUI.
  * Incremental build problems with our AWS library.
* Minor odd jobs related to the release & testing.

**Comment by Frizi** [comment]
- Stabilized list editor implementation, implemented restarting propagation of previously stopped events. With that, added support for dropdowns or other interactive widgets inside list editor. https://github.com/enso-org/enso/pull/6540
![image](https://user-images.githubusercontent.com/919491/236849717-ce6649c0-4932-44d3-8771-245ae71babed.png)

- Started working on applying new design for nodes, no visual progress yet: https://github.com/enso-org/enso/issues/6552
  - Read the code that manages color assignment to groups in component browser, as we want to match node backgrounds with them. Turns out the assigned colors aren't always the same, will need to figure out what to do about it.

**Comment by kazcw** [comment]
I've been working on reducing draw calls to improve render performance. In my current branch I have a 23% reduction. By the end of the week I plan to reach 40%. During this week I also plan to finish improving CB focus.

![newplot(3)](https://user-images.githubusercontent.com/1047859/236850159-5984111f-27ac-4127-b5e0-90c6db81fa25.png)

After this PR is finished, I'll benchmark it on various hardware and check how draw call count relates to render performance; they should be closely related, so I expect this change will make Enso smooth on hardware where it's currently marginal.
